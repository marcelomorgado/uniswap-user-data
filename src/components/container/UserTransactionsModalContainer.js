import React from "react";
import PropTypes from "prop-types";
import UserTransactionsModal from "../presentational/UserTransactionsModal";
import { Query } from "react-apollo";
import { gql } from "apollo-boost";

const ITEMS_PER_PAGE = 15;

const GET_USER_TRANSACTIONS = gql`
  query Transaction($userId: String!, $itemsPerPage: Int) {
    transactions(first: $itemsPerPage, where: { user: $userId }) {
      id
      tx
      event
      block
      timestamp
      exchangeAddress
      tokenAddress
      tokenSymbol
      user
      ethAmount
      tokenAmount
      fee
    }
  }
`;

const GET_MORE_USER_TRANSACTIONS = gql`
  query Transaction($userId: String!, $cursor: String, $itemsPerPage: Int) {
    transactions(
      first: $itemsPerPage
      where: { user: $userId, id_lt: $cursor }
    ) {
      id
      tx
      event
      block
      timestamp
      exchangeAddress
      tokenAddress
      tokenSymbol
      user
      ethAmount
      tokenAmount
      fee
    }
  }
`;

class UserTransactionsModalContainer extends React.PureComponent {
  state = {
    hasNextPage: true,
    isNextPageLoading: false,
  };

  render() {
    const { userId, open, handleClose } = this.props;
    const { hasNextPage, isNextPageLoading } = this.state;

    if (!userId) return <></>;

    return (
      <Query
        query={GET_USER_TRANSACTIONS}
        variables={{ userId, itemsPerPage: ITEMS_PER_PAGE }}
      >
        {({ loading, error, data, fetchMore }) => {
          if (loading) return <p>Loading...</p>;
          if (error) return <p>Error :(</p>;

          const { transactions } = data;

          if (!transactions.length) {
            this.setState({ hasNextPage: false });
          }

          const onLoadMore = () => {
            const lastTransaction = transactions[transactions.length - 1];
            const { id: cursor } = lastTransaction;

            this.setState({ isNextPageLoading: true });

            const updateQuery = (prevResult, { fetchMoreResult: newData }) => {
              const { transactions: prevTransactions } = prevResult;
              const { transactions: newTransactions } = newData;
              const hasNewData = newTransactions.length;
              let data;

              if (!hasNewData) {
                data = prevResult;
                this.setState({ hasNextPage: false });
              } else {
                const transactions = [...prevTransactions, ...newTransactions];
                data = { transactions };
              }

              this.setState({ isNextPageLoading: false });
              return data;
            };

            return fetchMore({
              query: GET_MORE_USER_TRANSACTIONS,
              variables: {
                userId,
                cursor,
                itemsPerPage: ITEMS_PER_PAGE,
              },
              updateQuery,
            });
          };

          return (
            <UserTransactionsModal
              open={open}
              handleClose={handleClose}
              userId={userId}
              transactions={transactions}
              onLoadMore={onLoadMore}
              hasNextPage={hasNextPage}
              isNextPageLoading={isNextPageLoading}
            />
          );
        }}
      </Query>
    );
  }
}

UserTransactionsModalContainer.propTypes = {
  open: PropTypes.bool.isRequired,
  handleOpen: PropTypes.func.isRequired,
  handleClose: PropTypes.func.isRequired,
  userId: PropTypes.string,
};

export default UserTransactionsModalContainer;
