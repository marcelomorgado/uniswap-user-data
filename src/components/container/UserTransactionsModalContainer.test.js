import React from "react";
import { shallow } from "enzyme";
import UserTransactionsModalContainer from "./UserTransactionsModalContainer";

it("renders the component", () => {
  const props = {
    open: true,
    handleClose: () => {},
    handleOpen: () => {},
    userId: "0x123User",
  };

  expect(
    shallow(<UserTransactionsModalContainer {...props} />)
  ).toMatchSnapshot();
});
