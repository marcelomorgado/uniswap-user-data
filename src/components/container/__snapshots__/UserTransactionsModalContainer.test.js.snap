// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
<Query
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "Transaction",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "first",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "itemsPerPage",
                      },
                    },
                  },
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "where",
                    },
                    "value": Object {
                      "fields": Array [
                        Object {
                          "kind": "ObjectField",
                          "name": Object {
                            "kind": "Name",
                            "value": "user",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                          },
                        },
                      ],
                      "kind": "ObjectValue",
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "transactions",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "tx",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "event",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "tokenAddress",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "tokenSymbol",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "user",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "ethAmount",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "tokenAmount",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "userId",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "Int",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "itemsPerPage",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 251,
        "start": 0,
      },
    }
  }
  variables={
    Object {
      "itemsPerPage": 15,
      "userId": "0x123User",
    }
  }
>
  <Component />
</Query>
`;
